Source/Core/VideoBackends/Metal/MTLStateTracker.mm

```diff
@@ -0,0 +1,938 @@
+{
+  m_usage.clear();
+  m_size = new_size;
+  m_pos = 0;
+}
+
+// MARK: - StateTracker
+
+Metal::StateTracker::StateTracker() : m_backref(std::make_shared<Backref>(this))
+{
+  m_flags.should_apply_label = true;
+  m_fence = MRCTransfer([g_device newFence]);
+  for (MRCOwned<MTLRenderPassDescriptor*>& rpdesc : m_render_pass_desc)
+  {
+    rpdesc = MRCTransfer([MTLRenderPassDescriptor new]);
+    [[rpdesc depthAttachment] setStoreAction:MTLStoreActionStore];
+    [[rpdesc stencilAttachment] setStoreAction:MTLStoreActionStore];
+  }
+  m_resolve_pass_desc = MRCTransfer([MTLRenderPassDescriptor new]);
+  auto color0 = [[m_resolve_pass_desc colorAttachments] objectAtIndexedSubscript:0];
+  [color0 setLoadAction:MTLLoadActionLoad];
+  [color0 setStoreAction:MTLStoreActionMultisampleResolve];
+  MTLTextureDescriptor* texdesc =
+      [MTLTextureDescriptor texture2DDescriptorWithPixelFormat:MTLPixelFormatRGBA8Unorm
+                                                         width:1
+                                                        height:1
+                                                     mipmapped:NO];
+  [texdesc setTextureType:MTLTextureType2DArray];
+  [texdesc setUsage:MTLTextureUsageShaderRead];
+  [texdesc setStorageMode:MTLStorageModePrivate];
+  m_dummy_texture = MRCTransfer([g_device newTextureWithDescriptor:texdesc]);
+  [m_dummy_texture setLabel:@"Dummy Texture"];
+  for (size_t i = 0; i < std::size(m_state.samplers); ++i)
+  {
+    SetSamplerForce(i, RenderState::GetLinearSamplerState());
+    SetTexture(i, m_dummy_texture);
+  }
+}
+
+Metal::StateTracker::~StateTracker()
+{
+  FlushEncoders();
+  std::lock_guard<std::mutex> lock(m_backref->mtx);
+  m_backref->state_tracker = nullptr;
+}
+
+// MARK: BufferPair Ops
+
+Metal::StateTracker::Map Metal::StateTracker::AllocateForTextureUpload(size_t amt)
+{
+  amt = (amt + 15) & ~15ull;
+  CPUBuffer& buffer = m_texture_upload_buffer;
+  u64 last_draw = m_last_finished_draw.load(std::memory_order_acquire);
+  bool needs_new = buffer.usage.PrepareForAllocation(last_draw, amt);
+  if (__builtin_expect(needs_new, false))
+  {
+    // Orphan buffer
+    size_t newsize = std::max<size_t>(buffer.usage.Size() * 2, 4096);
+    while (newsize < amt)
+      newsize *= 2;
+    MTLResourceOptions options =
+        MTLResourceStorageModeShared | MTLResourceCPUCacheModeWriteCombined;
+    buffer.mtlbuffer = MRCTransfer([g_device newBufferWithLength:newsize options:options]);
+    [buffer.mtlbuffer setLabel:@"Texture Upload Buffer"];
+    ASSERT_MSG(VIDEO, buffer.mtlbuffer, "Failed to allocate MTLBuffer (out of memory?)");
+    buffer.buffer = [buffer.mtlbuffer contents];
+    buffer.usage.Reset(newsize);
+  }
+
+  size_t pos = buffer.usage.Allocate(m_current_draw, amt);
+
+  Map ret = {buffer.mtlbuffer, pos, reinterpret_cast<char*>(buffer.buffer) + pos};
+  assert(pos <= buffer.usage.Size() &&
+         "Previous code should have guaranteed there was enough space");
```

Comment:
This should use Dolphin's `ASSERT` or `ASSERT_MSG` macros (actually, it looks like there are several places that use plain `assert` instead of those). Unless there's some reason why that's incompatible with objective-C or something?